### PR TITLE
Better error message and tests for bad await

### PR DESCRIPTION
**Pull request by huantianad**
Currently attempting to use `await` in a non-`async` context gives the error message `'yield' only allowed in an iterator`.
This pull request changes `await` to raise a better error by first checking if the template is used in an `async` proc, and raising an error if it is not.

### CODE REVIEW

File: lib/pure/asyncmacro.nim [hunk 1]
```diff
@@ -121,14 +121,36 @@ proc verifyReturnType(typeName: string, node: NimNode = nil) =
...
 
 template await*[T](f: Future[T]): auto {.used.} =
   var internalTmpFuture: FutureBase = f
-  yield internalTmpFuture
-  (cast[typeof(f)](internalTmpFuture)).read()
+  when isInAsync(internalTmpFuture):
+    yield internalTmpFuture
+    (cast[typeof(f)](internalTmpFuture)).read()
+  else:
+    static: error "Can only 'await' inside a 'async' marked proc."
```
Comment (konsumlamm) [inline lib/pure/asyncmacro.nim]:
```suggestion
    static: error "Can only 'await' inside a proc marked as 'async'."
```

Comment (huantianad) [reply lib/pure/asyncmacro.nim]: Should we be even more specific on the error message? Maybe mention "proc marked with the 'async' pragma" or is that unnecessarily verbose.

Comment (beef331) [reply lib/pure/asyncmacro.nim]: Perhaps the message should be along the lines of `"Can only 'await' inside a proc marked as 'async'. If wanting to call a 'async' proc in non-async scope use 'waitFor'"`

**Comment by Araq** [comment]
This seems excessive. We should look into if we can improve the compiler's error message directly.

**Comment by huantianad** [comment]
So we would overwrite the error handler for this specific situation of `yield`ing a FutureBase outside a valid iterator? How hard would that be?

Edit: another simpler alternative is to use a `compiles()` to check if `yield` will compile, which might work just as well, i.e.
```nim
template await*[T](f: Future[T]): auto {.used.} =
  template yieldFuture = yield FutureBase()

  when compiles(yieldFuture):
    var internalTmpFuture: FutureBase = f
    yield internalTmpFuture
    (cast[typeof(f)](internalTmpFuture)).read()
  else:
    static: error(
      "Can only 'await' inside a proc marked as 'async'. Use " &
      "'waitFor' when calling an 'async' proc in a non-async scope instead")
```

**Comment by Araq** [comment]
Oh, that's a good idea too. Please see if it works.

**Comment by huantianad** [comment]
Seems to work well, can't think of any edge cases, other than the obvious await iterator of `FutureBase`.

I'll need to fix nimout looking for `/` instead of `\` in windows tests, and maybe consider a better error message for multisync.

**Comment by huantianad** [comment]
Looked into it a bit and I'm not sure how to fix the path test error when using nimout, other than just making separate windows and Linux tests. Anything I missed?

File: lib/pure/asyncmacro.nim [hunk 2]
```diff
@@ -126,9 +126,16 @@ template await*(f: typed): untyped {.used.} =
...
+  template yieldFuture = yield FutureBase()
+
+  when compiles(yieldFuture):
+    var internalTmpFuture: FutureBase = f
+    yield internalTmpFuture
+    (cast[typeof(f)](internalTmpFuture)).read()
+  else:
+    static: error(
+      "Can only 'await' inside a proc marked as 'async'. Use " &
+      "'waitFor' when calling an 'async' proc in a non-async scope instead")
```
Comment (ringabout) [inline lib/pure/asyncmacro.nim]: Try something like below to get better error message report
```nim
macro errorAsync[T](yieldFuture: Future[T]) =
  error(
        "Can only 'await' inside a proc marked as 'async'. Use " &
        "'waitFor' when calling an 'async' proc in a non-async scope instead",
        yieldFuture)

template await*[T](f: Future[T]): auto {.used.} =
  template yieldFuture = yield FutureBase()

  when compiles(yieldFuture):
    var internalTmpFuture: FutureBase = f
    yield internalTmpFuture
    (cast[typeof(f)](internalTmpFuture)).read()
  else:
    errorAsync(f)
```

> test.nim(68, 8) Error: Can only 'await' inside a proc marked as 'async'. Use 'waitFor' when calling an 'async' proc in a non-async scope instead

Instead of

`../../lib/pure/asyncmacro.nim(136, 18) Error: Can only 'await' inside a proc marked as 'async'. Use 'waitFor' when calling an 'async' proc in a non-async scope instead`

Comment (huantianad) [reply lib/pure/asyncmacro.nim]: Oh I forgot about that, that's much nicer! if only there was a way to get the NimNode for the entire template so IDEs like VSCode would highlight the entire `await` statement.

**Comment by huantianad** [comment]
Think that wraps up most of the loose ends, anything else to work on @Araq @xflywind? If not I can do a squash and rebase.

**Review by ringabout — COMMENTED** [review]
LGTM

**Comment by Araq** [comment]
In retrospect I would have simply left the error message as is. Can't hurt to learn how the programming language that you use actually works.
